lua/autorun/client/cl_devtools.lua

```diff
@@ -1,144 +1,171 @@
+ 
+-- lua/STRUCT/*
 
-SWAMP_DEV.structTypes = {
+SWAMP_DEV.structEnvironments = {
     ["weapons"] = function(curClass)
-        SWEP = weapons.GetStored(curClass)
+        local found = weapons.GetStored(curClass)
+        return found and {SWEP=found} or nil
     end,
     ["entities"] = function(curClass)
-        ENT = scripted_ents.GetStored(curClass)
+        local found = scripted_ents.GetStored(curClass)
+        return found and {ENT=found} or nil
     end,
     ["effects"] = function(curClass)
-        local fxTab = effects.GetList()
-        local fx = "effects/" .. curClass
-
-        for _, v in ipairs(fxTab) do
-            if (v.Folder == fx) then
-                EFFECT = v
-
-                return
+        local allEffects = effects.GetList()
+        local targetEffect = "effects/" .. curClass
+
+        for _, effect in ipairs(allEffects) do
+            if effect.Folder == targetEffect then
+                return {
+                    EFFECT = effect
+                }
             end
         end
     end
 }
 
--- Returns a string to print
-local function refreshFile(path, filename)
-    local code = file.Read(path, "MOD")
-
-    -- current filename as class format
-    if SWAMP_DEV.itemStruct then
-        local parentFolder = string.gsub(string.sub(path, 1, -#filename - 2), ".*/", "")
 
-        -- if parent isn't a struct then use parent as class, else use filename
-        if (not SWAMP_DEV.structTypes[parentFolder]) then
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](parentFolder)
-        else
-            SWAMP_DEV.structTypes[SWAMP_DEV.itemStruct](string.TrimRight(string.gsub(path, "^.*/", ""), ".lua"))
+local ROOT_DIRECTORY = "addons/contrib/"
+local refreshFolder
+local loadFile
+do 
+    local loadFunc = function() end
+    function loadFile(filePath, environmentVars)
+        local code = file.Read(filePath, "MOD")
+        if not code then 
+            error("No file exists with the file path: " .. filePath)
         end
-    end
-
-    BaseGamemode = engine.ActiveGamemode()
-    GM = baseclass.Get(BaseGamemode)
-
-    -- handle include
-    for s in string.gmatch(code, "include%s*%b()") do
-        local reqInclude = string.sub(string.gsub(s, "[%s%(%)\'\"]+", ""), 8)
-        local includePath = string.sub(path, 1, -#filename - 1) .. reqInclude
-
-        -- First check the parent folder
-        if (file.Exists(includePath, "MOD")) then
-            print("INCLUDING: " .. refreshFile(includePath, string.gsub(reqInclude, ".*/", "")))
-            -- Then check the lua/ game folder
-        elseif (file.Exists("addons/contrib/lua/" .. reqInclude, "MOD")) then
-            print("INCLUDING: " .. refreshFile("addons/contrib/lua/" .. reqInclude, string.gsub(reqInclude, ".*/", "")))
-        else
-            print("Include handled by file")
+        environmentVars = environmentVars or {}
+       
+        -- setup our environment
+        local environment = _G
```

Comment:
clean this up?

Reply:
How so?

Reply:
why are you changing the name of this variable instead of just using _G itself? and don't say "for optimization"

Reply:
Its for performance as is everything i do. I focus on optimising everything as much as possible no mattter its calling context.

Reply:
Okay, but, this kind of "optimization" is totally missing the forest because of the trees.

First of all this isn't in a PrePlayerDraw hook. This code only runs _when a developer is using the devtools_ and _they have just edited a file_. That happens probably less than 100 times per day on average, and happens only when the developer is editing something, so we really don't need to give them the extra 0.00001% frames per second this kind of edit would realize (if you generously estimate that this saves 100ns and say they are updating one file every second). Clearly this doesn't make any significant difference, especially compared to something like PrePlayerDraw, which I estimate probably gets run a cumulative 20 billion times per day by our players.

At the same time though this addon _does_ have a significant performance issue in it. The problem is how when checking the file modification times, you check all of the files at once every two seconds rather than amortizing those checks over the two seconds. This is going to spike the frame time every time you check it. I wouldn't be surprised if the spike is quite significant as well, because I assume file.Time requires a synchronous disk read which might take something like 1ms (which adds up over 400 files). I don't notice any issue myself but I have a nvme SSD which is pretty much the best case scenario for that.

I understand the desire to make everything "as optimal as possible", but this is the type of optimization that should probably only be done by a compiler/transpiler if it's wanted. It seems fairly simple to statically analyze a program and localize all repeated global variable accesses, and you might actually get a measurable improvement when doing this kind of optimization across all of the code and in more impactful places. I did this optimization in the gamemode's animations.lua, which is full of hooks that run for every player every tick, and improved the server tickrate by something like 10%

So this change obviously doesn't have any significant upside in performance, but why do I dislike it? It's bloat - it makes the code more confusing. When you see _G getting set, you know you're setting globals, but when you see "environment" getting set, it takes extra analysis to figure out that "environment" is actually just globals. It makes the underlying program/AST more complicated and makes the file longer and take longer to read. Obviously for just this singular change, the negative impact is minor, but what it demonstrates is that you're not trying to make the code as simple as possible. Efficiency doesn't just mean absolute maximum performance when the margin of improvement is so insignificant; optimizing your program to be short and easier to understand is a more important objective.

Reply:
I understand your response and will remember that for the future.